dockerize app (#9)
pathwaycom/llm-app: README.md

| --------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `contextful`    | In this mode, the app will index the documents located in the `data/pathway-docs` directory. These indexed documents are then taken into account when processing queries. The pathway pipeline being run in this mode is located at `llm_app/pathway_pipelines/contextful/app.py`. |
| `contextful_s3` | This mode operates similarly to the contextful mode. The main difference is that the documents are stored and indexed from an S3 bucket, allowing the handling of a larger volume of documents. This can be more suitable for production environments.                             |
| `contextless`   | This pipeline calls OpenAI ChatGPT API but does not use an index when processing queries. It relies solely on the given user query.                                                                                                                                                |

> | `local`         | This mode runs the application using Huggingface Transformers, which eliminates the need for external APIs. It provides a convenient way to use state-of-the-art NLP models locally without additional services.                                                                   |

Reviewer: ```suggestion
| `local`         | This mode runs the application using Huggingface Transformers, which eliminates the need for the data to leave the machine. It provides a convenient way to use state-of-the-art NLP models locally without additional services.                                                                   |
```
or "the need for external services" and then remove "without additional services"?
I would say that HF can be thought of as a kind of an external API (I know it doesn't send any data, but is not created by pathway).

    | PATHWAY_REST_CONNECTOR_HOST | Specifies the host IP for the REST connector in Pathway. For the dockerized version, set it to `0.0.0.0` Natively, you can use `127.0.0.1`. interfaces.                                                                                                    |
    | PATHWAY_REST_CONNECTOR_PORT | Specifies the port number on which the REST connector service of the Pathway should listen. Here, it is set to 8080.                                                                                                                                       |
    | OPENAI_API_TOKEN            | The API token for accessing OpenAI services. If you are not running the local version, please remember to replace it with your personal API token, which you can generate from your account on [openai.com](https://platform.openai.com/account/api-keys). |
    | PATHWAY_CACHE_DIR           | Specifies the directory where cache is stored. You could use /tmp/cache.                                                                                                                                                                                   |
    

    ```bash

>     PIPELINE_MODE=contextful

Reviewer: I think I'd prefer mode as an argument to the script (as earlier). IMHO, env vars are good for settings that do not alter the logic much and it applies to all other values but the mode changes the behavior completely.
However, you can disagree.

Author: Let's do it in a separate pr

    | PATHWAY_CACHE_DIR           | Specifies the directory where cache is stored. You could use /tmp/cache.                                                                                                                                                                                   |
    

    ```bash
    PIPELINE_MODE=contextful
    PATHWAY_REST_CONNECTOR_HOST=0.0.0.0
    PATHWAY_REST_CONNECTOR_PORT=8080
    OPENAI_API_TOKEN=<Your Token>
    PATHWAY_CACHE_DIR=/tmp/cache
    ```

#### Using Docker:

Docker is a tool designed to make it easier to create, deploy, and run applications by using containers. Here is how to use Docker to build and run the LLM App:

  - **Build and Run with Docker** The first step is to build the Docker image for the LLM App. You do this with the docker build command.
    Build the image:
    ```bash
    docker build -t llm-app .
    ```
    After your image is built, you can run it as a container. You use the docker run command to do this
    ```bash

>     docker run -it -p 8080:8080 -v ./data/:/app/data/ llm-app

Reviewer: ```suggestion
    docker run -it -p 8080:8080 llm-app
```

>     cd llm_app/

Reviewer: ```suggestion
    cd llm-app/
```
I didn't notice it earlier. It should be the name of the repo.

    pip install --upgrade --extra-index-url https://packages.pathway.com/966431ef6ba -r requirements.txt
    ```

- **Run the App:** You can start the application with the command:
    ```bash 
    cd llm_app/
    python main.py
    ```

### Usage

1. **Send REST queries** (in a separate terminal window): These are examples of how to interact with the application once it's running. `curl` is a command-line tool used to send data using various network protocols. Here, it's being used to send HTTP requests to the application.
    ```bash
    curl --data '{"user": "user", "query": "How to connect to Kafka in Pathway?"}' http://localhost:8080/ | jq

    curl --data '{"user": "user", "query": "How to use LLMs in Pathway?"}' http://localhost:8080/ | jq
    ```
    Please change `localhost` to `0.0.0.0` if you are running the app on docker.

> 2. **Test reactivity by adding a new file:** This shows how to test the application's ability to react to changes in data by adding a new file and sending a query.

Reviewer: It won't work in docker. It'd be better to add a data volume than put in a Dockerfile. Putting it in Dockerfile gives users less flexibility for experimentation.
